satellite/metainfo/loop.go

```diff
@@ -0,0 +1,247 @@
+	defer mon.Task()(&ctx)(&err)
+
+	loop.someCtx = ctx
+	// if we remove these two lines the test passes
+	ctx, cancel := context.WithCancel(ctx)
+	loop.cancel = cancel
+	defer close(loop.done)
+
+	for {
+		err := loop.runOnce(ctx)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+// runOnce goes through metainfo one time and sends information to observers
+func (loop *Loop) runOnce(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	var observers []*observerContext
+
+	defer func() {
+		for _, observer := range observers {
+			observer.Finish()
+		}
+	}()
+
+	// wait for the first observer, or exit because context is canceled
+	select {
+	case observer := <-loop.join:
+		observers = append(observers, observer)
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	// after the first observer is found, set timer for CoalesceDuration and add any observers that try to join before the timer is up
+	timer := time.NewTimer(loop.config.CoalesceDuration)
+waitformore:
+	for {
+		select {
+		case observer := <-loop.join:
+			observers = append(observers, observer)
+		case <-timer.C:
+			break waitformore
+		case <-ctx.Done():
+			for _, observer := range observers {
+				observer.HandleError(ctx.Err())
+			}
+			observers = nil
+			return ctx.Err()
+		}
+	}
+
+	err = loop.metainfo.Iterate(ctx, "", "", true, false,
+		func(ctx context.Context, it storage.Iterator) error {
+			var item storage.ListItem
+
+			// iterate over every segment in metainfo
+			for it.Next(ctx, &item) {
+				fmt.Println("iterating...")
+				pointer := &pb.Pointer{}
+
+				err = proto.Unmarshal(item.Value, pointer)
+				if err != nil {
+					// TODO: figure out what to do
+					// return LoopError.New("error unmarshalling pointer %s", err)
+					continue
+				}
+
+				path := item.Key.String()
+				pathElements := storj.SplitPath(path)
+				isLastSeg := len(pathElements) >= 2 && pathElements[1] == "l"
+
+				nextObservers := observers[:0]
+
+				// send segment info to every observer
+				for _, observer := range observers {
+					remote := pointer.GetRemote()
+					if remote != nil {
+						if observer.HandleError(observer.RemoteSegment(ctx, path, pointer)) {
+							continue
+						}
+
+						if isLastSeg {
+							if observer.HandleError(observer.RemoteObject(ctx, path, pointer)) {
+								continue
+							}
+						}
+					} else if observer.HandleError(observer.InlineSegment(ctx, path, pointer)) {
+						continue
+					}
+
+					select {
+					case <-observer.ctx.Done():
+						observer.HandleError(observer.ctx.Err())
+						continue
+					default:
+					}
+
+					// for the next segment, only iterate over observers that did not have an error or canceled context
+					nextObservers = append(nextObservers, observer)
+				}
+
+				observers = nextObservers
+				if len(observers) == 0 {
+					return nil
+				}
+
+				// if context has been canceled, send the error to observers and exit. Otherwise, continue
+				select {
+				case <-loop.someCtx.Done():
```

Comment:
This shouldn't be here, the cancellation is propagated.

Reply:
I added it because I couldn't figure out why `<-ctx.Done()` was not happening. It was to experiment and see if the parent context was canceled. Which it was. But the check below never happens, even when the parent context is canceled. This is why I am confused.

Reply:
Things in satellite are implemented such that they don't call close before run finishes. I suspect that's what is causing you problems.

Reply:
well in the test, we are creating a new loop, separate from the satellite, that way we can pass in a cancelable context we create within the test. The test in question isn't even calling `Close()`. It is just canceling the context passed into `Run()`

Reply:
When we run the test with the code as is right now, we do see "the original context is done" printed out. When we get rid of `loop.someCtx` and its associated `case` here, and re run the test, we just see "continuing iteration" and the loop never exits due to a context cancelation

Reply:
If we remove the two lines in `Run` that create a new `ctx` with cancel based on the old context, the test passes as expected. It is only when we create a new context with cancel that it fails